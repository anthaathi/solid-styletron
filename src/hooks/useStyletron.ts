import { useContext } from 'solid-js';
import {
  StyletronContext,
  StyletronTheme,
  StyletronThemeContext,
} from '../contexts/styletron';
import type { StyleObject } from 'styletron-standard';

export function useStyletron(): [
  css: (input: StyleObject | StyleObject[]) => string,
  $theme: StyletronTheme
] {
  const styletron = useContext(StyletronContext);
  const $theme = useContext(StyletronThemeContext);

  if (!styletron) {
    throw new Error('No styletron provided');
  }

  return [
    (input) => {
      if (Array.isArray(input)) {
        const prevValue = {};

        input.forEach((obj) => {
          mergeDeep(prevValue, obj);
        });

        return styletron?.renderStyle(prevValue);
      }

      return styletron?.renderStyle(input);
    },
    $theme,
  ];
}

/**
 * Performs a deep merge of `source` into `target`.
 * Mutates `target` only but not its objects and arrays.
 *
 * @author inspired by [jhildenbiddle](https://stackoverflow.com/a/48218209).
 */
function mergeDeep(target: object, source: object) {
  const isObject = (obj: object) => obj && typeof obj === 'object';

  if (!isObject(target) || !isObject(source)) {
    return source;
  }

  Object.keys(source).forEach((key) => {
    // @ts-ignore
    const targetValue = target[key];
    // @ts-ignore
    const sourceValue = source[key];

    if (Array.isArray(targetValue) && Array.isArray(sourceValue)) {
      // @ts-ignore
      target[key] = targetValue.concat(sourceValue);
    } else if (isObject(targetValue) && isObject(sourceValue)) {
      // @ts-ignore
      target[key] = mergeDeep(Object.assign({}, targetValue), sourceValue);
    } else {
      // @ts-ignore
      target[key] = sourceValue;
    }
  });

  return target;
}
