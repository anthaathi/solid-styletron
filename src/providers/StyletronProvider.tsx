import type { Component, JSX } from 'solid-js';
import type { StandardEngine } from 'styletron-standard';
import {
  StyletronContext,
  StyletronTheme,
  StyletronThemeContext,
} from '../contexts/styletron';

export interface StyletronProviderProps {
  client: StandardEngine;
  theme: StyletronTheme;
  children: JSX.Element;
}

export const StyletronProvider: Component<StyletronProviderProps> = (props) => {
  return (
    <StyletronContext.Provider value={props.client}>
      <StyletronThemeContext.Provider value={props.theme}>
        {props.children}
      </StyletronThemeContext.Provider>
    </StyletronContext.Provider>
  );
};
