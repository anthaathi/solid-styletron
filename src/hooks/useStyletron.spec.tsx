import { render } from 'solid-testing-library';
import { StyletronProvider } from '../providers/StyletronProvider';
import { Server } from 'styletron-engine-atomic';
import { useStyletron } from './useStyletron';

describe('useStyletron', () => {
  it('should ', function () {
    const Dl = () => {
      const [css] = useStyletron();

      return (
        <div>
          <div
            class={css([
              { 'marginLeft': '12px', ':hover': { top: '12px' } },
              { 'marginRight': '12px', ':hover': { padding: '12px' } },
            ])}
          />
          <div class={css({ padding: '12px' })}></div>
        </div>
      );
    };

    const client = new Server();

    const ele = render(() => (
      <StyletronProvider client={client} theme={{}}>
        <Dl />
        {client.getStylesheets().map((style) => {
          return <style innerHTML={style.css} />;
        })}
      </StyletronProvider>
    ));

    expect(ele.container).toMatchSnapshot();
  });
});
