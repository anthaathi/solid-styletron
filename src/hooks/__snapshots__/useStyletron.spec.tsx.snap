// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStyletron should  1`] = `
<div>
  <div>
    <div
      class="ae af ag ah"
    />
    <div
      class="ai"
    />
  </div>
  <style>
    .ae{margin-left:12px}.af:hover{top:12px}.ag:hover{padding:12px}.ah{margin-right:12px}.ai{padding:12px}
  </style>
</div>
`;
